### PR TITLE
Make it possible for track selectors to include timing cuts (via a ref)

### PR DESCRIPTION
**Pull request by lgray**
Adds in the capability for making timing cuts in track selector modules via a ref.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

File: CommonTools/RecoAlgos/interface/RecoTrackSelector.h [hunk 1]
```diff
@@ -25,10 +26,12 @@ class RecoTrackSelector: public RecoTrackSelectorBase {
     init(event,es);
     selected_.clear();
     for( reco::TrackCollection::const_iterator trk = c->begin();
-         trk != c->end(); ++ trk )
-      if ( operator()(*trk) ) {
+         trk != c->end(); ++ trk ) {
+      reference_type tkref(c,std::distance(c->begin(),trk));
```
Comment (slava77) [inline CommonTools/RecoAlgos/interface/RecoTrackSelector.h]: what's the cost of this cycle to make a ref and then dereference it (in order to pass the same collection pointer to the method downstream)?

**Comment by lgray** [comment]
Event times in the validator were not very different. So, marginal.

**Comment by slava77** [comment]
On 11/8/16 6:22 AM, Lindsey Gray wrote:
> Event times in the validator were not very different. So, marginal.

which validator?
If these selectors are used in HLT, even 1ms per module can be too
expensive.

**Comment by lgray** [comment]
MultiTrackValidator

**Comment by lgray** [comment]
A quick check reveals these selectors are not used in the HLT. Unless I am missing something.

**Comment by lgray** [comment]
@slava77 where does this one live in the queue?

**Comment by slava77** [comment]
On 11/12/16 5:27 PM, Lindsey Gray wrote:
> @slava77 <https://github.com/slava77> where does this one live in the queue?
> 

It's with Carl

**Comment by lgray** [comment]
ok, thanks!

**Comment by cvuosalo** [comment]
Timing tests of this PR are in progress.

**Comment by monttj** [comment]
+1

**Comment by cvuosalo** [comment]
Tests of workflow 25202.0_TTbar_13 with 150 events against baseline CMSSW_8_1_X_2016-11-13-0000 show some time increase for modules affected by the changes in this PR. In two test runs (with and without DQM/Validation), the affected modules show an increase in CPU time of around 25%. The affected modules contribute very little to the total time, and total time per event doesn't change significantly.

Test with DQM and Validation:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
  +0.235769      +0.00%         0.77 ms/ev ->         0.97 ms/ev trackWithVertexRefSelector
  +0.231492      +0.00%         0.81 ms/ev ->         1.02 ms/ev trackWithVertexRefSelectorBeforeSorting
  +0.239171      +0.00%         0.73 ms/ev ->         0.93 ms/ev generalTracksFromPV
  ---------- ------------     --------                  ----       ------------
```
Another measure with DQM & Validation:
```
        trackWithVertexRefSelectorBeforeSorting          1.39921 ms/ev -> 1.70389 ms/ev
        trackWithVertexRefSelector                       1.3264 ms/ev -> 1.62408 ms/ev
```


Test without DQM or Validation:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +0.322756      +0.00%         0.74 ms/ev ->         1.03 ms/ev trackWithVertexRefSelector
   +0.297348      +0.00%         0.80 ms/ev ->         1.08 ms/ev trackWithVertexRefSelectorBeforeSorting
  ---------- ------------     --------                  ----       ------------
```

**Comment by cvuosalo** [comment]
+1

For #16500 f7ec74391204c1bad0271a0f5423fc711e44549a

Adding capability to make timing cuts in track selector modules via a reference.

The code changes are satisfactory, and Jenkins tests show no significant differences. A test of workflow 25202.0_TTbar_13 with 150 events against baseline CMSSW_8_1_X_2016-11-13-0000 also shows no significant differences. CPU time does increase for modules affected by this PR (see above), but the increase is acceptable given the very small proportion these modules contribute to overall processing time.

In addition to the timing results listed above, here are timing differences for additional modules affected by this PR. They show no increase in CPU time. The measurement was made as described above with DQM and Validation for workflow 25202.0
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   -0.087638      -0.00%        15.70 ms/ev ->        14.38 ms/ev generalTracksPt09
   -0.034253      -0.00%        13.99 ms/ev ->        13.52 ms/ev probeTracks
   -0.104694      -0.00%        11.23 ms/ev ->        10.11 ms/ev cutsRecoTracksInitialStep
   -0.090783      -0.00%        19.07 ms/ev ->        17.42 ms/ev cutsRecoTracksHp
   -0.086563      -0.00%         7.22 ms/ev ->         6.62 ms/ev cutsRecoTracksPt09Hp
   -0.085985      -0.00%         5.36 ms/ev ->         4.92 ms/ev cutsRecoTracksDetachedTripletStep
   -0.085538      -0.00%         9.56 ms/ev ->         8.78 ms/ev cutsRecoTracksInitialStepHp
   -0.077553      -0.00%        14.28 ms/ev ->        13.21 ms/ev cutsRecoTracksBtvLike
   -0.076033      -0.00%         5.23 ms/ev ->         4.85 ms/ev cutsRecoTracksPixelPairStepHp
   -0.069502      -0.00%         5.53 ms/ev ->         5.16 ms/ev cutsRecoTracksPixelPairStep
  ---------- ------------     --------                  ----       ------------
```

**Comment by davidlange6** [comment]
not needed in 81x. (phase2 dev will evolve in 90x)

**Pull request closed without merging**